staging/src/k8s.io/dynamic-resource-allocation/structured/pools.go

```diff
@@ -27,41 +27,77 @@ import (
 	draapi "k8s.io/dynamic-resource-allocation/api"
 )
 
+func NodeMatches(node *v1.Node, nodeNameToMatch string, allNodesMatch bool, nodeSelector *v1.NodeSelector) (bool, error) {
+	switch {
+	case nodeNameToMatch != "":
+		return node != nil && node.Name == nodeNameToMatch, nil
+	case allNodesMatch:
+		return true, nil
+	case nodeSelector != nil:
+		selector, err := nodeaffinity.NewNodeSelector(nodeSelector)
+		if err != nil {
+			return false, fmt.Errorf("failed to parse node selector %s: %w", nodeSelector.String(), err)
+		}
+		return selector.Match(node), nil
+	}
+
+	return false, nil
+}
+
 // GatherPools collects information about all resource pools which provide
 // devices that are accessible from the given node.
 //
 // Out-dated slices are silently ignored. Pools may be incomplete (not all
 // required slices available) or invalid (for example, device names not unique).
 // Both is recorded in the result.
-func GatherPools(ctx context.Context, slices []*resourceapi.ResourceSlice, node *v1.Node) ([]*Pool, error) {
+func GatherPools(ctx context.Context, slices []*resourceapi.ResourceSlice, node *v1.Node, features Features) ([]*Pool, error) {
 	pools := make(map[PoolID]*Pool)
-	nodeName := ""
-	if node != nil {
-		nodeName = node.Name
-	}
 
 	for _, slice := range slices {
+		if !features.PartitionableDevices && (len(slice.Spec.SharedCounters) > 0 || slice.Spec.PerDeviceNodeSelection != nil) {
+			return nil, fmt.Errorf("DRAPartitionableDevices disabled with SharedCounters or PerDeviceNodeSelection set in ResourceSlice")
```

Comment:
Returning an error means that all scheduling of pods involving DRA stops until the DRA driver publishing this slice gets downgraded and these slices get removed or replaced.

Is that what we want? We could instead ignore unsupported slices.

If we want an error, then include the ResourceSlice and driver name to give admins a chance to figure out what it is complaining about? Ultimately this should be reported as a pod scheduling failure (to be verified?).